packages/teams-js/src/internal/videoPerformanceStatistics.ts

```diff
@@ -0,0 +1,128 @@
+  /**
+   * The duration in milliseconds that the data were collected
+   */
+  duration: number;
+  /**
+   * The number of frames that were processed in the duration
+   */
+  sampleCount: number;
+  /**
+   * An array that presents counts of frames that were finished in n milliseconds:
+   * distributionBins[frameProcessingDurationInMs]=frameCount.
+   * For example, distributionBins[10] = 5 means that 5 frames were processed in 10 milliseconds.
+   */
+  distributionBins: Uint32Array;
+};
+
+export class VideoPerformanceStatistics {
+  private static readonly initialSessionTimeoutInMs = 1000;
+  private static readonly maxSessionTimeoutInMs = 1000 * 30;
+
+  private currentSession: {
+    startedAtInMs: number;
+    timeoutInMs: number;
+    effectId: string;
+    frameWidth: number;
+    frameHeight: number;
+  };
+
+  private frameProcessingStartedAt: number;
+  private distributionBins: Uint32Array;
+  private sampleCount = 0;
+  private timeoutId: string;
+
+  public constructor(
+    distributionBinSize: number,
+    /**
+     * Function to report the statistics result
+     */
+    private reportStatisticsResult: (result: VideoPerformanceStatisticsResult) => void,
+  ) {
+    this.distributionBins = new Uint32Array(distributionBinSize);
+  }
+
+  /**
+   * Call this function before processing every frame
+   */
+  public processStarts(effectId: string, frameWidth: number, frameHeight: number): void {
+    VideoFrameTick.tick();
+    if (!this.suitableForThisSession(effectId, frameWidth, frameHeight)) {
+      this.reportAndResetSession(this.getStatistics(), effectId, frameWidth, frameHeight);
+    }
+    this.start();
+  }
+
+  public processEnds(): void {
+    // calculate duration of the process and record it
+    const durationInMillisecond = performance.now() - this.frameProcessingStartedAt;
+    const binIndex = Math.floor(Math.max(0, Math.min(this.distributionBins.length - 1, durationInMillisecond)));
+    this.distributionBins[binIndex] += 1;
+    this.sampleCount += 1;
+  }
+
+  private getStatistics(): VideoPerformanceStatisticsResult {
+    if (!this.currentSession) {
+      return null;
+    }
+    return {
+      effectId: this.currentSession.effectId,
+      frameHeight: this.currentSession.frameHeight,
+      frameWidth: this.currentSession.frameWidth,
+      duration: performance.now() - this.currentSession.startedAtInMs,
+      sampleCount: this.sampleCount,
+      distributionBins: this.distributionBins.slice(),
+    };
+  }
+
+  private start(): void {
+    this.frameProcessingStartedAt = performance.now();
+  }
+
+  private suitableForThisSession(effectId: string, frameWidth: number, frameHeight: number): boolean {
+    return (
+      this.currentSession &&
+      this.currentSession.effectId === effectId &&
+      this.currentSession.frameWidth === frameWidth &&
+      this.currentSession.frameHeight === frameHeight
+    );
+  }
+
+  private reportAndResetSession(result, effectId, frameWidth, frameHeight): void {
```

Comment:
When the effect param change, should we consider it as a new session?

Reply:
Yes, updated